log4j-core/src/main/java/org/apache/logging/log4j/core/util/DefaultShutdownCallbackRegistry.java

```diff
@@ -87,34 +86,33 @@ public void run() {
     }
 
     private static class RegisteredCancellable implements Cancellable {
-        // use a reference to prevent memory leaks
-        private final Reference<Runnable> hook;
+        private Runnable callback;
         private Collection<Cancellable> registered;
 
         RegisteredCancellable(final Runnable callback, final Collection<Cancellable> registered) {
+            this.callback = callback;
             this.registered = registered;
-            hook = new SoftReference<>(callback);
```

Comment:
I'm not sure why this used a `SoftReference` rather than `WeakReference` since the former is notorious for GC problems, and I don't think we want to hold the callback reference until the heap is otherwise full.